### ISSUE TITLE
Pjax and inline scripts

### ISSUE THREAD

**Issue by llfm** [yiisoft/yii2#6655]
When Pjax retrieves some `<script>` tags and there's inline scripts after this, the script code isn't executed (async). This [issue](https://github.com/defunkt/jquery-pjax/issues/331) offers a workaround/alternative to this kind of problem. 

Maybe the yii2-pjax should follow this, as the framework has a good amount of inline script.

**Comment by lynicidn** [comment]
better way add relation in AppAssets  to all scripts

**Comment by nkovacs** [comment]
I have a fix for this: https://github.com/defunkt/jquery-pjax/issues/331#issuecomment-98824483
Like I said in that comment, this way you lose parallel downloads, so this might not be the best solution for pjax, and might not be accepted into upstream. But Yii uses the pattern of registering a script file and then a small inline initializer script a lot, so you could merge this into your fork.
Let's wait for a response from the pjax devs first.

**Comment by mikehaertl** [comment]
@nkovacs Shouldn't `executeScriptTags()` also be moved? In my case where I load for example a GriedView with pjax (using current Yii 2 Pjax asset and manual pjax invocation with `$.pjax()`) execution stops at `context.html(container.contents)` with an error like `yii.gridView() is not a function`. So the script stops before `executeScriptTags()` is even called and thus the `<script>` tag for `yii.gridView.js` can never be handled.

**Comment by mikehaertl** [comment]
@samdark 
Slightly related to this issue: I found, that the script filter in `yii.js:initScriptFitler` would not work with PJax, because PJax doesn't seemt to use `$.ajax()` so the `$.ajaxPrefilter` is never called. So even with @nkovacs Pjax fix merged - and even if that fix avoids loading of duplicate scripts inside Pjax - there's still danger that scripts are loaded twice. E.g. a script may get loaded with the main page, then later again from a Pjax call. Please correct me if I'm wrong here.

But if the my assumption is right, I think the script filter needs some enhancement. Worth a new issue?

**Comment by nkovacs** [comment]
The code in pjax does its own matching to prevent double loading.

That error you described shouldn't be happening. container.contents shouldn't contain script tags. 
This change here removes all script tags so that they can be executed serially later:
https://github.com/nkovacs/jquery-pjax/commit/b54d08273b520334c591c37bdd4ee07296157d1e#diff-256b2e7d3ffe31142c00fa086e15d3a6L765

**Comment by nkovacs** [comment]
Also, it's important to note that inline scripts are always reloaded, so they should not affect the global state - e.g. they must not contain `$(document).on('click', '.some-class', function() { ... })`, because that event handler will never be removed (you can work around it using pjax's events though).
They should only contain code like `$('.some-element-inside-pjax-container').on('click', '.some-class', function() { ... })`, because when that element is removed, the event handlers are also removed.

Conversely, script files are loaded only once, so they cannot contain code that binds directly to elements inside the pjax container, because on a pjax "reload" the script won't run again.

**Comment by mikehaertl** [comment]
Thanks, makes sense.

> The code in pjax does its own matching to prevent double loading.

Hmm, but what if you combine `$.ajax()` with `$.pjax()` on a page? In the first case, Yii's script filter will take care of not loading duplicate scripts. Will Pjax take note, when a new script is loaded outside of Pjax?

**Comment by mikehaertl** [comment]
After studying more code I think I can refine my above statement:

With @nkovacs patch, pjax doesn't use jQuery's script loading mechanism. It basically inserts a `<script>` tag into the header to load the script. Thus the script filter in `$.ajaxPrefilter` is not called and the script file loaded from pjax is not inserted into yii's list of loaded scripts.

So if after a Pjax request a `$.ajax()` request is made, which receives the same `<script src="..">` in its response, Yii's script filter will not prevent it from being loaded again. That is because the script filter only queries the `loadedScripts` once in `initScriptFilter()`. It will not know about Pjax having inserted that script already.

To fix this, we could rebuild the `loadedScripts` from `$('script[src]')` inside the `$.ajaxPrefilter` handler each time. Or at least merge it with any new scripts in the DOM that are not in `loadedScripts` yet.

Opinions?

**Comment by nkovacs** [comment]
Jquery's script insertion (.html and friends) is bad, because it's synchronous. My patch to pjax is asynchronous, except for inline scripts, but I should fix that too (Chrome already warns about deprecation).

Unfortunately, this is where pjax being an external plugin that has different design goals becomes a problem. This issue is not a priority for them, so it doesn't look like my patch is going to be included. My patch also forces all scripts to load serially, which is necessary for correctness, but it also means you lose parallel script downloads. Most of the time, you can get away with loading scripts in parallel, but when you have an external script that depends on another one before it, and both are loaded with pjax, you'll get a race condition without my patch. Since it's not possible to tell what depends on what, everything has to be completely serial.

> To fix this, we could rebuild the loadedScripts from $('script[src]') inside the $.ajaxPrefilter handler each time. Or at least merge it with any new scripts in the DOM that are not in loadedScripts yet.

The former is probably a bad idea, because your script tags might disappear when you remove the whole content from the html, so you'll re-run the script.

My patch always appends to head to prevent this problem. The idea is that external script files are libraries, and once you load them, they're loaded and can't be unloaded, and there's no need to load them again.

**Comment by mikehaertl** [comment]
> The idea is that external script files are libraries, and once you load them, they're loaded and can't be unloaded, and there's no need to load them again.

Right, but that's exactly the problem. I'm not sure if you're aware of Yii's solution for this. It's in `yii.js` here:

https://github.com/yiisoft/yii2/blob/89eb3dff144cffeb1b72107a235afc345ac2320e/framework/assets/yii.js#L308

So please read again what I wrote above. I still think, that there's a potential problem in Yii's and Pjax' way to prevent duplicate loading of scripts. Yii will not know that Pjax has already loaded a script, if you use `$.ajax()` after some `$.pjax()` call.

**Comment by SilverFire** [comment]
Duplicates #3680 
Duplicates #4814
Duplicates #8540 
Duplicates #8916

**Comment by nkovacs** [comment]
I don't think #8540 is the same problem. This is about the order of execution.

**Comment by SilverFire** [comment]
I really want to fix this Pjax bunch of :hankey:. I am going to try your [commit](https://github.com/nkovacs/jquery-pjax/commit/b54d08273b520334c591c37bdd4ee07296157d1e) because it will probably fix these bugs. When you want to take apart - feel free to contact me in https://gitter.im/silverfire

**Comment by nkovacs** [comment]
@mikehaertl Yes, there is a problem. I avoid inserting scripts via $.ajax() because of this, and only use pjax.

**Comment by mikehaertl** [comment]
@nkovacs Actually I think the script filter as implemented in `yii.js` is the only reliable solution. Only the client can know, which scripts are already loaded, so filtering should happen there. I wonder if we could convince the Pjax guys to fire an event, each time a `<script>` tag is removed from a Pjax response. This way Yii could listen to that event and use its own mechanism (including script filter) to load that script (or not).

**Comment by SilverFire** [comment]
I think this will never happen. It will be easier to maintain fork, that pushing Pjax guys to make changes.

**Comment by nkovacs** [comment]
I think the way yii.js prevents duplicate script files won't work with pjax. Yii.js uses an ajax prefilter for script files, but that only work if you use $.html to insert the script tag (which causes jquery to fire of a synchronous ajax call for the script file, and that triggers the prefilter). That doesn't work with my patch (as well as the original pjax implementation), since that inserts a script tag to avoid the synchronous ajax call.

**Comment by SilverFire** [comment]
Since now JavaScripts load through PJAX will be processed by `jQuery.ajaxPrefiler`

https://github.com/yiisoft/jquery-pjax/commit/7a47df7b36d1031fbd576929e040ebab588b207b
